avalanche/NEW_core.py

```diff
@@ -0,0 +1,226 @@
+        pass
+
+    def after_backward(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called after `criterion.backward()` by the `BaseTemplate`."""
+        pass
+
+    def after_training_iteration(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called after the end of a training iteration by the
+        `BaseTemplate`."""
+        pass
+
+    def before_update(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called before `optimizer.update()` by the `BaseTemplate`."""
+        pass
+
+    def after_update(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called after `optimizer.update()` by the `BaseTemplate`."""
+        pass
+
+    def after_training_epoch(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called after `train_epoch` by the `BaseTemplate`."""
+        pass
+
+    def before_eval_iteration(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called before the start of a training iteration by the
+        `BaseTemplate`."""
+        pass
+
+    def before_eval_forward(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called before `model.forward()` by the `BaseTemplate`."""
+        pass
+
+    def after_eval_forward(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called after `model.forward()` by the `BaseTemplate`."""
+        pass
+
+    def after_eval_iteration(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called after the end of an iteration by the
+        `BaseTemplate`."""
+        pass
+
+    def before_train_dataset_adaptation(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called before `train_dataset_adapatation` by the `BaseTemplate`."""
+        pass
+
+    def after_train_dataset_adaptation(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called after `train_dataset_adapatation` by the `BaseTemplate`."""
+        pass
+
+    def before_eval_dataset_adaptation(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called before `eval_dataset_adaptation` by the `BaseTemplate`."""
+        pass
+
+    def after_eval_dataset_adaptation(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called after `eval_dataset_adaptation` by the `BaseTemplate`."""
+        pass
+
+    # ====================================================================> NEW
+
+    def before_inner_updates(
+        self, strategy: Template, *args, **kwargs
+    ) -> CallbackResult:
+        """Called before `_inner_updates` by the `BaseTemplate`."""
+        pass
+
+    def inner_updates(
```

Comment:
That's true, but the reason I added the `inner_updates` callback as a plugin trigger is that, unlike supervised strategies, we don't have any "Naive" type of updates that I can set as default for meta-learning-based strategies.
 More precisely, in supervised strategies, we have the `training_epoch` function that is implemented in its most basic form (which is Naive fine-tuning), and you can augment it by adding new plugins. We don't have such a general structure for `training_epoch` inner updates, and it can be completely different from method to method. That's why I added it as a plugin trigger that has to be implemented by the user. Do you have other suggestions?